experimental/ActionPolyRing/src/Content.jl

```diff
@@ -0,0 +1,1373 @@
+
+__jtu_idx(dpr::DifferencePolyRing) = dpr.jet_to_upoly_idx
+
+__are_perms_up_to_date(dpr::DifferencePolyRing) = dpr.are_perms_up_to_date
+
+function __perm_for_sort(dpr::DifferencePolyRing)
+  if __are_perms_up_to_date(dpr) && isdefined(dpr, :permutation)
+    return dpr.permutation
+  end
+  __set_perm_for_sort!(dpr)
+  return dpr.permutation
+end
+
+function __perm_for_sort_poly(dpre::DifferencePolyRingElem)
+  dpr = parent(dpre)
+  if __are_perms_up_to_date(dpr) && isdefined(dpre, :permutation)
+    return dpre.permutation
+  end
+  __set_perm_for_sort!(dpr)
+  __set_perm_for_sort_poly!(dpre)
+  return dpre.permutation
+end
+
+#Setters for internals
+function __set_are_perms_up_to_date!(dpr::DifferencePolyRing, update::Bool)
+  dpr.are_perms_up_to_date = update
+end
+
+function __set_perm_for_sort!(dpr::DifferencePolyRing)
+  dpr.permutation = sortperm(dpr.(gens(__upr(dpr))); rev = true)
+  __set_are_perms_up_to_date!(dpr, true)
+end
+
+#If this is called, we assume that are_perms_up_to_date == true
+function __set_perm_for_sort_poly!(dpre::DifferencePolyRingElem)
+  __update_internals!(dpre)
+  dpre.permutation = sortperm(map(expo -> expo[parent(dpre).permutation], collect(exponents(data(dpre)))); rev = true)
+end
+
+### Differential ###
+data(dpre::DifferentialPolyRingElem) = dpre.p
+
+__upr(dpr::DifferentialPolyRing) = dpr.upoly_ring
+
+__jtv(dpr::DifferentialPolyRing) = dpr.jet_to_var
+
+__vtj(dpr::DifferentialPolyRing) = dpr.var_to_jet
+
+__jtu_idx(dpr::DifferentialPolyRing) = dpr.jet_to_upoly_idx
+
+__are_perms_up_to_date(dpr::DifferentialPolyRing) = dpr.are_perms_up_to_date
+
+function __perm_for_sort(dpr::DifferentialPolyRing)
+  if __are_perms_up_to_date(dpr) && isdefined(dpr, :permutation)
+    return dpr.permutation
+  end
+  __set_perm_for_sort!(dpr)
+  return dpr.permutation
+end
+
+function __perm_for_sort_poly(dpre::DifferentialPolyRingElem)
+  dpr = parent(dpre)
+  if __are_perms_up_to_date(dpr) && isdefined(dpre, :permutation)
+    return dpre.permutation
+  end
+  __set_perm_for_sort!(dpr)
+  __set_perm_for_sort_poly!(dpre)
+  return dpre.permutation
+end
+
+#Setters for internals
+function __set_are_perms_up_to_date!(dpr::DifferentialPolyRing, update::Bool)
+  dpr.are_perms_up_to_date = update
+end
+
+function __set_perm_for_sort!(dpr::DifferentialPolyRing)
+  dpr.permutation = sortperm(dpr.(gens(__upr(dpr))); rev = true)
+  __set_are_perms_up_to_date!(dpr, true)
+end
+
+#If this is called, we assume that are_perms_up_to_date == true
+function __set_perm_for_sort_poly!(dpre::DifferentialPolyRingElem)
+  __update_internals!(dpre)
+  dpre.permutation = sortperm(map(expo -> expo[parent(dpre).permutation], collect(exponents(data(dpre)))); rev = true)
+end
+
+#Check if the jet_to_var dictionary of apr could contain the key (i,jet).
+__is_valid_jet(apr::ActionPolyRing, i::Int, jet::Vector{Int}) = i in 1:nelementary_symbols(apr) && length(jet) == ndiffs(apr) && all(j -> j >= 0, jet)
```

Comment:
```suggestion
__is_valid_jet(apr::ActionPolyRing, i::Int, jet::Vector{Int}) = i in 1:nelementary_symbols(apr) && length(jet) == ndiffs(apr) && all(>=(0), jet)
```
can be slightly more efficient